### DISCUSSION TITLE
Bash, starship and bash-preexec

### DISCUSSION THREAD

**Discussion by devmatteini** [jdx/mise#1517] [Q&A]
Hi! I just started trying out `mise` and so far so good!

The only minor problem I found is the bash integration with [bash-preexec](https://github.com/rcaloras/bash-preexec/) and [starship](https://github.com/starship/starship/).

---

I installed with `mise` node 18 and 20 as global version.
Given this minimal `.bashrc`:
```bash
source ~/.bash-preexec.sh
eval "$(mise activate bash)"
eval "$(starship init bash)"
```

When I `cd` into a directory with a `.nvmrc` with `v18` starship displays `v20.11.0` as node version instead of `v18.19.0`.
After I run a new command starship re-draw the prompt and the correct version is shown.

If you remove `source ~/.bash-preexec.sh`, `mise activate` works as expected.
I was wondering if any of you had any idea what could be causing this issue when used with `bash-preexec`.

If I use shims, this issue is fixed.
```bash
source ~/.bash-preexec.sh
PATH="$HOME/.local/share/mise/shims:$PATH"
eval "$(starship init bash)"
```

There's only a ~5ms difference running node with shims, so it's not a big issue.

**Comment by Marukome0743** [comment]
This appears to be resolved in current mise versions.

Since [PR #3373](https://github.com/jdx/mise/pull/3373), included in mise v2024.12.1, Bash activation installs a directory-change hook that runs `hook-env` immediately after `cd`, before the next prompt is rendered.

I retested the original ordering on current main:

```bash
source ~/.bash-preexec.sh
eval "$(mise activate bash)"
# followed by the prompt integration
```

Using two directories with different mise-managed environment values, the value changed immediately after each `cd` and was already correct when the `PROMPT_COMMAND` callbacks ran. `bash-preexec` remained installed in the same session.

This should prevent Starship from seeing the previous directory's Node environment for the first prompt. Shims remain a valid alternative, but they should no longer be required as a workaround for this one-prompt delay.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
